# TFM: Parte de preprocesamiento de datos.
### Autores: 
- ####  Adrian
- ####  Carlos Mendívil Gómez
- ####  Guillermo
- ####  Jose

In [ ]:
# --- Análisis de Datos y Computación Numérica ---
import pandas as pd
import numpy as np

# --- Visualización de Datos ---
import seaborn as sns
import matplotlib.pyplot as plt

# --- Modelos Estadísticos ---
import statsmodels.api as sm

# --- Machine Learning con Scikit-Learn ---
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.cluster import KMeans
from sklearn.impute import SimpleImputer
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import OneHotEncoder, StandardScaler

# --- Utilidades y Librerías Estándar ---
import math
import re
import requests
import zipfile
from io import StringIO
from IPython.display import Markdown, display
import warnings

warnings.filterwarnings("ignore")

In [2]:
url = "https://raw.githubusercontent.com/guille1006/TFM/refs/heads/main/data2/total_data.csv"
df = pd.read_csv(url)

# Descripción general de las variables

<table>
  <tr><th>Campo</th><th>Variable valida</th><th>Tipo</th><th>Descripción</th></tr>
  
  <tr><td colspan="4" align="center"><b> Dirección del inmueble</b></td></tr>
  <tr><td>latitude</td><td>Si</td><td>float</td><td>Coordenada de latitud</td></tr>
  <tr><td>longitude</td><td>Si</td><td>float</td><td>Coordenada de longitud</td></tr>
  <tr><td>showAddress</td><td>Si</td><td>bool</td><td>Muestra la dirección o no</td></tr>
  <tr><td>address</td><td>Si</td><td>category</td><td>Calle, número y país del inmueble</td></tr>
  <tr><td>neighborhood</td><td>Si</td><td>category</td><td>Vecindario donde está ubicado el inmueble</td></tr>
  <tr><td>district</td><td>Si</td><td>category</td><td>Distrito donde está ubicado el inmueble</td></tr>
  <tr><td>municipality</td><td>Si</td><td>category</td><td>Municipio donde está ubicado el inmueble</td></tr>
  <tr><td>province</td><td>No</td><td>category</td><td>Provincia donde está ubicado el inmueble</td></tr>
  <tr><td>country</td><td>No</td><td>category</td><td>País donde está ubicado el inmueble</td></tr>

  <tr><td colspan="4" align="center"><b> Características propias del inmueble</b></td></tr>
  <tr><td>rooms</td><td>Si</td><td>int</td><td>Número de habitaciones del inmueble</td></tr>
  <tr><td>size</td><td>Si</td><td>float</td><td>Metros cuadrados del inmueble</td></tr>
  <tr><td>bathrooms</td><td>Si</td><td>int</td><td>Número de baños</td></tr>
  <tr><td>exterior</td><td>Si</td><td>bool</td><td>Tiene parte exterior o no</td></tr>
  <tr><td>status</td><td>Si</td><td>category</td><td>Estado de la vivienda</td></tr>
  <tr><td>detailedType.subTypology</td><td>Si</td><td>category</td><td>Subtipo de inmueble</td></tr>
  <tr><td>detailedType.typology</td><td>Si</td><td>category</td><td>Tipo de inmueble</td></tr>
  <tr><td>floor</td><td>Si</td><td>int</td><td>Altura del piso</td></tr>
  <tr><td>hasLift</td><td>Si</td><td>bool</td><td>Tiene ascensor o no</td></tr>
  <tr><td>parkingSpace.hasParkingSpace</td><td>Si</td><td>bool</td><td>Tiene espacio para aparcar o no</td></tr>
  <tr><td>parkingSpace.isParkingSpaceIncludedInPrice</td><td>Si</td><td>bool</td><td>Está el precio del aparcamiento incluido en el precio o no</td></tr>
  <tr><td>parkingSpace.parkingSpacePrice</td><td>Si</td><td>float</td><td>Precio del espacio de aparcamiento</td></tr>
  <tr><td>price</td><td>Si</td><td>float</td><td>Precio del inmueble</td></tr>
  <tr><td>priceInfo.price.amount</td><td>No</td><td>float</td><td>Precio del inmueble</td></tr>
  <tr><td>priceByArea</td><td>Si</td><td>float</td><td>Precio del metro cuadrado del inmueble</td></tr>
  <tr><td>priceInfo.price.currencySuffix</td><td>No</td><td>category</td><td>Moneda del precio</td></tr>
  <tr><td>priceInfo.price.priceDropInfo.formerPrice</td><td>Si</td><td>float</td><td>Precio anterior del inmueble</td></tr>
  <tr><td>priceInfo.price.priceDropInfo.priceDropPercentage</td><td>Si</td><td>float</td><td>Porcentaje de rebaja del precio con respecto al precio anterior</td></tr>
  <tr><td>priceInfo.price.priceDropInfo.priceDropValue</td><td>Si</td><td>float</td><td>Cantidad del dinero que se ha rebajado con respecto al precio anterior</td></tr>
  <tr><td>newDevelopment</td><td>Si</td><td>bool</td><td>Es vivienda de nueva construcción o no</td></tr>
  <tr><td>newDevelopmentFinished</td><td>Si</td><td>bool</td><td>Es vivienda de nueva construcción terminada o no</td></tr>
  <tr><td>topNewDevelopment</td><td>No</td><td>bool</td><td>Es un inmueble de nueva construcción o no</td></tr>
  <tr><td>highlight.groupDescription</td><td>Si</td><td>category</td><td></td></tr>

  <tr><td colspan="4" align="center"><b> Información relacionada con el anuncio</b></td></tr>
  <tr><td>description</td><td>Si</td><td>category</td><td>Descripción del anuncio, escrita por el dueño o la inmobiliaria</td></tr>
  <tr><td>numPhotos</td><td>Si</td><td>int</td><td>Número de fotos</td></tr>
  <tr><td>has360</td><td>Si</td><td>bool</td><td>Tiene imagen 360º del piso o no</td></tr>
  <tr><td>has3DTour</td><td>Si</td><td>bool</td><td>Tiene tour 3D o no</td></tr>
  <tr><td>hasVideo</td><td>Si</td><td>bool</td><td>Tiene video o no</td></tr>
  <tr><td>hasPlan</td><td>Si</td><td>bool</td><td>Tiene vista en plano o no</td></tr>
  <tr><td>hasStaging</td><td>Si</td><td>bool</td><td>Tiene visualización HomeStaging o no</td></tr>
  <tr><td>operation</td><td>No</td><td>category</td><td>Tipo de operación: Venta, alquiler...</td></tr>
  <tr><td>suggestedTexts.subtitle</td><td>Si</td><td>object</td><td>Subtítulo del anuncio</td></tr>
  <tr><td>suggestedTexts.title</td><td>Si</td><td>object</td><td>Título del anuncio</td></tr>

  <tr><td colspan="4" align="center"><b> Información propia de Idealista</b></td></tr>
  <tr><td>externalReference</td><td>No</td><td>object</td><td>Referencia externa usada por la página web de Idealista</td></tr>
  <tr><td>thumbnail</td><td>No</td><td>object</td><td>Enlace a una imagen del anuncio</td></tr>
  <tr><td>topPlus</td><td>Si</td><td>bool</td><td>Pertenece al plan Plus o no</td></tr>
  <tr><td>highlight.groupDescription</td><td>Si</td><td>category</td><td>Tipo de grupo de descripcion</td></tr>
  <tr><td>url</td><td>No</td><td>object</td><td>URL del anuncio</td></tr>
  <tr><td>notes</td><td>No</td><td>list</td><td>Lista vacía, no sabemos su uso</td></tr>
</table>


### Explicación de las variables no usadas:
- **Identificadores y Metadatos:** Variables como `url`, `externalReference` y `thumbnail`, que actúan como identificadores únicos o enlaces y no contienen información generalizable.

- **Variables Constantes:** Campos con un valor único para toda la muestra, como `country` (Spain), `province` (Madrid), `operation` (sale) y priceInfo.price.currencySuffix (EUR), que carecen de poder discriminatorio.

- **Variables Redundantes:** El campo `priceInfo.price.amount`, que es un duplicado exacto de la variable `price`.

- **Variables sin Varianza:** Columnas como `notes` y `TopNewDevelopment`, que se encontraron vacías, nulas o con un valor constante (False) en la totalidad de los registros.

# Limpieza de datos
El proceso de **limpieza de datos** se inicia con una fase de **análisis exploratorio preliminar**, cuyo objetivo principal es obtener una visión global del conjunto de datos almacenado en el *DataFrame*.  

In [3]:
# Vamos a mostrar una visión general del dataframe, para ver la distribución de las variables
print(df.shape)
df.describe().T

(26267, 49)


,count,mean,std,min,25%,50%,75%,max
bathrooms,26267.0,2.346442e+00,1.557807e+00,0.000000,1.000000e+00,2.000000e+00,3.000000e+00,2.900000e+01
latitude,26267.0,4.043244e+01,1.148277e-01,40.001899,4.038818e+01,4.042643e+01,4.046575e+01,4.110693e+01
longitude,26267.0,-3.717539e+00,1.638774e-01,-4.494274,-3.769397e+00,-3.698660e+00,-3.656073e+00,-3.071336e+00
numPhotos,26267.0,2.919210e+01,1.947732e+01,0.000000,1.600000e+01,2.600000e+01,3.900000e+01,1.900000e+02
parkingSpace.parkingSpacePrice,414.0,3.064735e+04,2.041348e+04,60.000000,1.500000e+04,2.800000e+04,4.000000e+04,1.000000e+05
price,26267.0,8.877257e+05,1.177656e+06,11000.000000,2.699450e+05,4.890000e+05,1.050000e+06,1.850000e+07
priceByArea,26267.0,5.173715e+03,3.526807e+03,2.000000,2.650000e+03,3.990000e+03,6.860000e+03,3.714300e+04
priceInfo.price.amount,26267.0,8.877257e+05,1.177656e+06,11000.000000,2.699450e+05,4.890000e+05,1.050000e+06,1.850000e+07
priceInfo.price.priceDropInfo.formerPrice,2806.0,6.127779e+05,5.973913e+05,32000.000000,2.500000e+05,3.990000e+05,7.500000e+05,7.500000e+06
priceInfo.price.priceDropInfo.priceDropPercentage,2806.0,4.614754e+00,3.116659e+00,1.000000,3.000000e+00,4.000000e+00,6.000000e+00,4.500000e+01


In [4]:
# Voy a realizar un estudio de las variables y del tipo de estas
info_types = pd.DataFrame(
    {"Tipo de columna": df.dtypes, 
     "Valores únicos": df.nunique(dropna=False),
     "Valores nulos": df.isnull().sum()}
).sort_values(by="Valores nulos", ascending=False)

info_types

,Tipo de columna,Valores únicos,Valores nulos
parkingSpace.parkingSpacePrice,float64,67,25853
newDevelopmentFinished,object,3,23536
priceInfo.price.priceDropInfo.formerPrice,float64,830,23461
priceInfo.price.priceDropInfo.priceDropValue,float64,276,23461
priceInfo.price.priceDropInfo.priceDropPercentage,float64,27,23461
detailedType.subTypology,object,14,17080
highlight.groupDescription,object,4,14883
parkingSpace.isParkingSpaceIncludedInPrice,object,3,14374
parkingSpace.hasParkingSpace,object,2,14374
neighborhood,object,198,10291


El siguiente paso en el proceso de depuración consistió en **descartar aquellas variables previamente identificadas como prescindibles**. 

De este modo, se garantiza que el *DataFrame* resultante contenga únicamente las variables útiles y pertinentes para los objetivos del estudio.  


In [ ]:
# Columnas que no son necesarias para la limpieza de datos inicial
cols_to_drop = [
    'url', 'externalReference', 'thumbnail', 'propertyCode', 
    'description', 'suggestedTexts.title', 'suggestedTexts.subtitle',
    'notes', 'address', 'country', 'province', 'operation', 
    'priceInfo.price.currencySuffix', 'priceInfo.price.amount',
    'priceInfo.price.priceDropInfo.priceDropValue', 'priceInfo.price.priceDropInfo.formerPrice',
    'topNewDevelopment', 'latitude', 'longitude'
]

df.drop(columns=cols_to_drop, inplace=True)

column_tot = df.columns

Otro aspecto fundamental dentro del proceso de limpieza de datos consiste en **identificar los posibles valores que puede adoptar cada variable categórica y realizar un conteo de su frecuencia de aparición**. Para ello usaremos la función de `tabla_valres_unicos`

In [ ]:
def tabla_valores_unicos(df):
    """
    Devuelve un DataFrame con la frecuencia absoluta y relativa de los valores únicos 
    en cada columna de `df`.

    Los valores nulos se muestran como "Nulo". El resultado tiene un MultiIndex 
    (columna, valor) y columnas:
        - "n": frecuencia absoluta
        - "%": frecuencia relativa (0 a 1)

    Args:
        df (pandas.DataFrame): DataFrame de entrada.

    Returns:
        pandas.DataFrame: Tabla de frecuencias por columna y valor único.
    """
    df_frec = pd.DataFrame()

    for col in df.columns:
        # Generare, solo para la visualización, que estos elementos se llamen "Nulo"
        series_fillna = df[col].fillna("Nulo")

        # Realizamos el conteo de cada valor y su frecuencia
        series_cont = series_fillna.value_counts()
        series_frec = series_cont/series_cont.sum()

        # Usaremos un DataFrame con MultiIndex para poder visualizar y acceder mejor a cada columna
        arrays_multiindex = [[col]* len(series_cont), series_cont.index.tolist()]
        multi_index = pd.MultiIndex.from_arrays(arrays_multiindex)

        result = pd.DataFrame({"n":series_cont.values, "%": series_frec.values}, index=multi_index)
        df_frec = pd.concat([df_frec, result], axis=0)
    
    return df_frec

In [ ]:
df_frec = tabla_valores_unicos(df)

## Tratamiento de los valores nulos


### Variables Booleanas nulas

| Variable                                 | Nº de Nulos | Tipo   | Estrategia de Imputación                   |
| :--------------------------------------- | :---------- | :----- | :----------------------------------------- |
| **exterior** | 7038        | object | Rellenar con `False` y convertir a binaria. |
| **hasLift** | 6716        | object | Rellenar con `False` y convertir a binaria. |
| **newDevelopmentFinished** | 23536       | object | Son aquellos que son de nueva obra, podemos poner el resto como `False` |
| **parkingSpace.hasParkingSpace** | 14374       | object | Solo tiene valores `True` y `NaN`, por lo que estos últimos serán `False` |
| **parkingSpace.isParkingSpaceIncludedInPrice**| 14374    | object | Nulos iguales a `hasParkingSpace`, repetir proceso. |

In [ ]:
# exterior
df["exterior"] = df["exterior"].fillna(False)

# hasLift
df["hasLift"] = df["hasLift"].fillna(False)

# newDevelopmentFinished	
df["newDevelopmentFinished"] = df["newDevelopmentFinished"].fillna(False)

# parkingSpace.hasparkingSpace
df["parkingSpace.hasParkingSpace"] = df["parkingSpace.hasParkingSpace"].fillna(False)

# parkingSpace.isParkingSpaceIncludedInPrice
df["parkingSpace.isParkingSpaceIncludedInPrice"] = df["parkingSpace.isParkingSpaceIncludedInPrice"].fillna(False)

### Variables catégoricas nulas

| Variable | Nº de Nulos | Tipo | Estrategia de Imputación |
| :--- | :--- | :--- | :--- |
| **floor** | 7207 | object | Se estandarizan los valores no numéricos (ej. 'bj', 'st') a sus equivalentes numéricos (0, -1).<br> Los nulos se tratan como planta baja (0) y se crea una variable `floor_cat` para capturar el efecto no lineal de estas categorías especiales. |
| **detailedType.subTypology** | 17080 | object | Se imputa con una nueva categoría, 'Vacío', dada que no existe subtipolgia en ese caso. |
| **district** | 3450 | object | Se imputan los nulos mediante geolocalización inversa, utilizando las coordenadas para asignar el distrito correspondiente. |
| **highlight.groupDescription** | 14883 | object | Se imputa con la categoría 'Vacio', asumiendo que la ausencia de valor implica la ausencia de una promoción. |
| **neighborhood** | 10291 | object | La columna se elimina debido a su alta cardinalidad y nulos, y se reemplaza por una feature binaria, `has_neighborhood`. |
| **status** | 597 | object | Se imputan los nulos como 'New development', basándose en la correlación observada con la variable `newDevelopment`, lo que resuelve una inconsistencia en los datos. |

In [ ]:
# floor
def floor_function(value):
    """ 
    Procesa la información del campo 'floor' de un inmueble y clasifica su valor.

    Args:
        value (pandas.Series or dict): Registro que debe contener:
            - "floor": valor de planta (string).
            - "detailedType.typology": tipología del inmueble.

    Returns:
        pandas.Series: Serie con dos elementos:
            - float: planta en formato numérico (0.0–15).
            - str: categoría asignada ("Numerica" o etiqueta original).
    """
    numeric_cat = "Numerica"
    x = value["floor"]

    len_x = len(x)

    numeric = [c for c in x if c.isdigit()]
    alnum = [c for c in x if not c.isalnum()]

    if len(numeric) == len_x:
        res_num = float(x)
        res_cat = numeric_cat

    elif alnum:
        if alnum[0] in [".", ","]:
            res_num = float(x)
        elif alnum[0] == "+":
            res_num = float("".join(numeric))
        elif alnum[0] == "-":
            res_num = float(x)
        res_cat = numeric_cat

    else:
        res_num = 0.0
        if value["detailedType.typology"] in ["chalet", "countryHouse"]:
            res_cat = numeric_cat
        else:
            res_cat = x

    if res_num > 15:
        res_num = 15
    return pd.Series([res_num, res_cat])

In [ ]:
# floor
df["floor"] = df["floor"].fillna("Nulo")
df_floor = df.apply(floor_function, axis=1)
df["floor"], df["floor_cat"] = df_floor[0], df_floor[1]

# detailedType.subTypology
df["detailedType.subTypology"] = df["detailedType.subTypology"].fillna("Vacío")

# highlight.groupDescription
df["highlight.groupDescription"] = df["highlight.groupDescription"].fillna("Vacío")

# status
df["status"] = df["status"].fillna("newdevelopment")

# Neighborhood
df["has_neighborhood"] = (~df["neighborhood"].isna()).astype(int)

### Variables númericas nulas
| Variable                                        | Nº de Nulos | Tipo    | Estrategia de Imputación o Transformación                                                              |
| :---------------------------------------------- | :---------- | :------ | :----------------------------------------------------------------------------------------------------- |
| **parkingSpace.parkingSpacePrice** | 25853       | float64 | Imputar nulos con 0, en relacion con la variable hasParkingSpace y isParkingSpaceIncludedInPrice. <br>En el caso de falta de informacion  imputamos por la mediana |
| **priceInfo.price.priceDropInfo.formerPrice** | 23461       | float64 | Se elimará debido a su redundancia, tenemos la misma informacion con price y la nueva columna de descuento.                                |
| **priceInfo.price.priceDropInfo.priceDropPercentage**| 23461| float64 | Imputar nulos con 0, asumiendo que la ausencia de valor indica que no ha habido rebaja de precio.      |
| **priceInfo.price.priceDropInfo.priceDropValue**| 23461       | float64 | Imputar nulos con 0, y crear una nueva variable binaria a partir de sus datos.      |

In [ ]:
# parkingSpace.parkingSpacePrice
df["parkingSpace.parkingSpacePrice"] = df["parkingSpace.parkingSpacePrice"].fillna(0.0)

# priceInfo.price.priceDropInfo.priceDropPercentage
df["priceInfo.price.priceDropInfo.priceDropPercentage"] = df["priceInfo.price.priceDropInfo.priceDropPercentage"].fillna(0.0)


# Diferenciacion de variables

In [7]:
from itertools import zip_longest
# variables booleanas
variable_bool = info_types[info_types["Valores únicos"]<3].index.to_list()

# variables numericas
variables_num = df.select_dtypes(include=["number"]).columns.tolist()

# variables categoricas
categoricas = []
for col in df.columns:
    if df[col].dtypes == "object":
        if df[col].nunique(dropna=False) <20:
            categoricas.append(col)
df[categoricas] = df[categoricas].astype("category")
variables_cat = df.select_dtypes(include=["category"]).columns.tolist()


print(f"{"variable_bool":50}{"variables_num":<50}{"variables_cat":<10}")
print("------------------------------------------------------------------------------------------------------------------------")
for variable in zip_longest(variable_bool, variables_num, variables_cat, fillvalue=""): 
    print(f"{variable[0]:50}{variable[1]:<45}{variable[2]:<10}")

variable_bool                                     variables_num                                     variables_cat
------------------------------------------------------------------------------------------------------------------------
topPlus                                           bathrooms                                    detailedType.subTypology
showAddress                                       floor                                        detailedType.typology
parkingSpace.isParkingSpaceIncludedInPrice        numPhotos                                    highlight.groupDescription
parkingSpace.hasParkingSpace                      parkingSpace.parkingSpacePrice               propertyType
newDevelopment                                    price                                        status    
newDevelopmentFinished                            priceByArea                                  floor_cat 
hasVideo                                          priceInfo.price.priceDropInfo.priceD

In [8]:
sum_var = 0
variables = variable_bool + variables_num + variables_cat


print(f"Columnas totales: {len(column_tot)}")
print(f"Variables: {len(variables)}")
print("Variables no usadas:")
print([x for x in column_tot if x not in variables])

Columnas totales: 31
Variables: 28
Variables no usadas:
['district', 'municipality', 'neighborhood']


## Estudio de las variables categoricas


| Variable                   | Tipo     | Estrategia de Tratamiento                                |
| :------------------------- | :------- | :------------------------------------------------------- |
| **detailedType.subTypology** | category | Agrupación de categorías poco representadas (< 0.65%).   |
| **detailedType.typology** | category | Distribución correcta.                                   |
| **district** | category | Agrupación de categorías poco representadas (< 0.065 %). |
| **highlight.groupDescription** | category | Distribución correcta.                                   |
| **municipality** | category | Agrupación de categorías poco representadas (< 0.5 %).   |
| **propertyType** | category | Distribución correcta.                                   |
| **status** | category | Distribución correcta.                                   |
| **floor_cat** | category | Cambios realizados anteriormente.     


In [ ]:
# Estudio variables categóricas
n = len(variables_cat)
ncols = 2
nrows = math.ceil(n / ncols)

fig, axes = plt.subplots(nrows, ncols, figsize=(14, 4 * nrows))
axes = axes.flatten()

for ax, col in zip(axes, variables_cat):
    sns.histplot(data=df, x=col, bins=20, ax=ax)
    ax.set_title(col)
for ax in axes[len(variables_cat):]:
    fig.delaxes(ax)

plt.tight_layout()
plt.show()

### Clustering de las variables geograficas

Un aspecto relevante en el análisis de la vivienda es el **barrio/ zona geografíca** en el que se ubica cada inmueble, pues determina o está correlacionado con múltiples factores importantes para el estudio: 
* Seguridad del barrio
* Transporte y movilidad
* Proximidad a puntos de interés (hospitales, centros educativos, etc.)
* Clase socioeconómica del barrio
* Distancia al centro de Madrid

Los datos recopilados originalmente están segmentados en un número excesivo de zonas geográficas, por lo que su uso directo como una única variable resulta poco fiable y poco informativo para los modelos. Para resolver esta fragmentación y obtener una variable más manejable y representativa, se optó por agrupar las zonas mediante técnicas de *clustering*.

En el *Jupyter Notebook* `TFM_Clustering` se describen con detalle los pasos metodológicos seguidos y la justificación para la elección de `k = 50`. En este documento solo se incluye la sección de código correspondiente a la fase de conclusión del proceso de agrupamiento.


In [ ]:
# Realización del Clustering basado en centroides (KMeans)

# Coordenadas de referencia (Puerta del Sol)
lat_centro, lon_centro = 40.416775, -3.703790

# Distancia euclidea a la Puerta del Sol
df["distancia_al_centro"] = np.sqrt((df["latitude"]-lat_centro)**2 + 
                                    (df["longitude"]-lon_centro)**2)

# Seleccionar y escalar variables para el clustering geográfico
geo_features = ["latitude", "longitude", "distancia_al_centro"]
X_geo = df[geo_features]

# Escalamos las variables 
scaler = StandardScaler()
X_geo_scaled = scaler.fit_transform(X_geo)

# Realizamos el KMeans
# Usamos un k_value de 50, gracias a la investigación previa 
k_value = 50
kmeans = KMeans(n_clusters=k_value, random_state=42, n_init="auto")
df["Zona_cluster"] = kmeans.fit_predict(X_geo_scaled)

### Onehot encoder de las categoricas

In [11]:
df_encoded = pd.get_dummies(df, columns=variables_cat, drop_first=True)

In [12]:
for col in df_encoded.columns:
    print(col)

bathrooms
district
exterior
floor
has360
has3DTour
hasLift
hasPlan
hasStaging
hasVideo
municipality
neighborhood
newDevelopment
newDevelopmentFinished
numPhotos
parkingSpace.hasParkingSpace
parkingSpace.isParkingSpaceIncludedInPrice
parkingSpace.parkingSpacePrice
price
priceByArea
priceInfo.price.priceDropInfo.priceDropPercentage
rooms
showAddress
size
topPlus
detailedType.subTypology_casaDePueblo
detailedType.subTypology_casale
detailedType.subTypology_caseron
detailedType.subTypology_cortijo
detailedType.subTypology_countryHouse
detailedType.subTypology_duplex
detailedType.subTypology_independantHouse
detailedType.subTypology_masia
detailedType.subTypology_palacio
detailedType.subTypology_penthouse
detailedType.subTypology_semidetachedHouse
detailedType.subTypology_studio
detailedType.subTypology_terracedHouse
detailedType.typology_countryHouse
detailedType.typology_flat
highlight.groupDescription_Top
highlight.groupDescription_Top+
highlight.groupDescription_Vacío
propertyType_count

In [13]:
len(df_encoded.columns)

56

In [14]:
df_encoded.to_csv("data2/data_fin.csv", index=False)

In [15]:
df.to_csv("data2/data_complete.csv", index=False)

## Estudio de las variables numéricas

| Variable                                 | Tipo      | Estrategia de Tratamiento                                                       |
| :--------------------------------------- | :-------- | :--------------------------------------------------------------------------------------------------------- |
| **floor** | Numérica  | Cambios ya realizados.|
| **price** | Numérica  | Detectar outliers según los dos criterios de la función e imputar nans.  |
| **size** | Numérica  | Detectar outliers según los dos criterios de la función e imputar nans.|
| **rooms** | Categórica| Cambio de int64 a category, tramificación para obtener una distribución más uniforme.                    |
| **bathrooms** | Categórica| Cambio de int64 a category, tramificación para obtener una distribución más uniforme.                    |
| **latitude** / **longitude** | Numérica  | Sin necesidad de cambios.      |
| **priceByArea** | Numérica  | Detectar outliers según los dos criterios de la función e imputar nans.                                      |
| **parkingSpace.parkingSpacePrice** | Numérica  | Sin necesidad de cambios.|
| **numPhotos** | Categórica| Cambio de int64 a category, tramificación para obtener una distribución más uniforme.                    |
| **showAddress** | Binaria   | Sin necesidad de cambios.                                                                            |
| **hasVideo** | Binaria   | Sin necesidad de cambios.                                                                            |
| **newDevelopment** | Binaria   | Sin necesidad de cambios.                                                                            |
| **newDevelopmentFinished** | Binaria   | Sin necesidad de cambios.                                                                            |
| **hasPlan** | Binaria   | Sin necesidad de cambios.                                                                            |
| **has3DTour** | Binaria   | Sin necesidad de cambios.                                                                            |
| **has360** | Binaria   | Sin necesidad de cambios.                                                                            |
| **hasStaging** | Binaria   | Clase minoritaria poco representada (< 0.05). Se evaluará su eliminación.                               |
| **topPlus** | Binaria   | Clase minoritaria poco representada (< 0.05). Se evaluará su eliminación.                               |
| **exterior** | Binaria   | Sin necesidad de cambios.                                                                            |
| **hasLift** | Binaria   | Sin necesidad de cambios.                                                                            |
| **parkingSpace.hasParkingSpace** | Binaria   | Sin necesidad de cambios.                                                                            |
| **parkingSpace.isParkingSpaceIncludedInPrice** | Binaria   | Sin necesidad de cambios.                                                                            |
| **has_neighborhood** | Binaria   | Feature de ingeniería ya creada. Se mantiene.                                                       |
| **has_Discount** | Binaria   | Feature de ingeniería ya creada. Se mantiene.

In [ ]:
n = len(variables_num)
fig, axes = plt.subplots(n, 1, figsize=(10, 2 * n), constrained_layout=True)

for ax, col in zip(axes, variables_num):
    sns.histplot(data=df, x=col, bins=20, ax=ax)
    ax.set_title(col)

plt.show()

# Tratamiento de Outliers

In [ ]:

def tratar_outliers(df, columnas_a_tratar, features_para_imputar):
    """
    Detecta outliers, los reemplaza por NaN y los imputa usando RandomForest,
    devolviendo un nuevo DataFrame limpio.
    """
    df_procesado = df.copy()

    # Detección y Reemplazo de Outliers
    for col in columnas_a_tratar:
        if abs(df_procesado[col].skew()) < 1:
            criterio1 = abs((df_procesado[col] - df_procesado[col].mean()) / df_procesado[col].std()) > 3
        else:
            mad = sm.robust.mad(df_procesado[col].dropna())
            criterio1 = abs((df_procesado[col] - df_procesado[col].median()) / mad) > 8

        q1, q3 = df_procesado[col].quantile([0.25, 0.75])
        criterio2 = ~df_procesado[col].between(q1 - 3 * (q3 - q1), q3 + 3 * (q3 - q1))
        df_procesado.loc[criterio1 & criterio2, col] = np.nan

    # Imputación de Nulos con RandomForest
    for col in columnas_a_tratar:
        if df_procesado[col].isnull().sum() == 0:
            continue

        # Preparar los datos para el modelo de imputación
        train = df_procesado[df_procesado[col].notnull()]
        test = df_procesado[df_procesado[col].isnull()]

        X_train = train[features_para_imputar].fillna(train[features_para_imputar].mean())
        y_train = train[col]
        X_test = test[features_para_imputar].fillna(train[features_para_imputar].mean())

        # Entrenar y predecir
        rf = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
        rf.fit(X_train, y_train)
        predictions = rf.predict(X_test)

        # Asignar las predicciones de vuelta al DataFrame
        df_procesado.loc[df_procesado[col].isnull(), col] = predictions

    return df_procesado

In [ ]:

features_para_imputar = [
    'size', 'latitude', 'longitude', 'floor', 'exterior', 'hasLift',
    'parkingSpace.hasParkingSpace', 'parkingSpace.isParkingSpaceIncludedInPrice',
    'parkingSpace.parkingSpacePrice', 'newDevelopment', 'newDevelopmentFinished',
    'has_Discount', 'has_neighborhood', 'hasVideo', 'hasPlan', 'has3DTour',
    'has360', 'hasStaging', 'topPlus', 'showAddress'
]
df = tratar_outliers(
    df,
    columnas_a_tratar=['price', 'priceByArea', 'size'],
    features_para_imputar=features_para_imputar
)

### Analísis descripitvo de las variables numericas
Ahora vamos a mostrar los estadisticos descriptivos de cada columna numerica para tener una idea de como se ha quedado nuestra base de datos

In [ ]:
# Calcular los descriptivos base
descriptivos_num = df[variables_num].describe().T

# Calcular estadísticos adicionales de forma eficiente con .agg()
additional_stats = df[variables_num].agg(['skew', 'kurt', lambda x: x.max() - x.min()]).T
additional_stats.rename(columns={'skew': 'Asimetria', 'kurt': 'Kurtosis', '<lambda>': 'Rango'}, inplace=True)

descriptivos_num = descriptivos_num.join(additional_stats)

# Mostrar la tabla de descriptivos completa
descriptivos_num